### PR TITLE
Convert to docker multistage Dockerfile

### PR DESCRIPTION
**Pull request by brancz**
As requested by the ART team, converting (our I believe last) Dockerfile to be multi-stage.

@metalmatze @squat @s-urbaniak @mxinden 

cc @sosiouxme @adammhaile

### CODE REVIEW

**Comment by brancz** [comment]
cc @pgier

File: Dockerfile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-FROM openshift/origin-base
+FROM openshift/origin-release:golang-1.10 AS builder
```
Comment (pgier) [inline Dockerfile]: I've been using the full path in other images
registry.svc.ci.openshift.org/openshift/release:golang-1.10 AS builder
But if it's fine to use just the relative path, that's probably better anyway.

File: Dockerfile [hunk 2]
```diff
@@ -7,12 +7,20 @@ RUN mkdir -p $GOBIN
 
 COPY . $GOPATH/src/github.com/grafana/grafana
```
Comment (pgier) [inline Dockerfile]: Pretty minor, but in the other images I've been setting the WORKDIR, and that would simplify later cd/cp commands.
```
WORKDIR /go/src/github.com/prometheus/prometheus
COPY . .
```

**Comment by pgier** [comment]
/hold
Just in case there are any other changes

**Comment by brancz** [comment]
We just chatted and this was indeed ready to be merged :slightly_smiling_face:

File: Dockerfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
-FROM openshift/origin-base
+FROM openshift/origin-release:golang-1.10 AS builder
 
 ENV GOPATH="/go"
```
Comment (sosiouxme) [inline Dockerfile]: Actually GOPATH in the builder includes /go and an SCL-specific path. I'm not certain what the consequences are of leaving that out. Thus I would suggest not setting this. I would use another variable or just "/go" (which is shorter anyway...)

File: Dockerfile [hunk 4]
```diff
@@ -7,12 +7,20 @@ RUN mkdir -p $GOBIN
...
+#RUN yum install -y make git
+RUN cd $GOPATH/src/github.com/grafana/grafana && go run build.go build
+RUN cp $GOPATH/src/github.com/grafana/grafana/bin/linux-$(go env GOARCH)/grafana-server /usr/bin/
+RUN rm -rf $GOPATH/src/github.com/grafana/grafana/.git
+
+FROM openshift/origin-base
+
+ENV GOPATH="/go"
+ENV GOBIN="${GOPATH}/bin"
+ENV PATH="${GOBIN}:${PATH}"
```
Comment (sosiouxme) [inline Dockerfile]: do you really want `/go/bin` in your path when you're not putting anything there?

File: Dockerfile [hunk 5]
```diff
@@ -7,12 +7,20 @@ RUN mkdir -p $GOBIN
...
+RUN mkdir -p $GOBIN
+
+COPY --from=builder /usr/bin/grafana-server /usr/bin/grafana-server
+COPY --from=builder $GOPATH/src/github.com/grafana/grafana $GOPATH/src/github.com/grafana/grafana
```
Comment (sosiouxme) [inline Dockerfile]: hmm, curious, why do you need the source code in the final image?

**Pull request merged**